tests/stub/authorization/test_user_switching.py

```diff
@@ -0,0 +1,378 @@
+            with self.session(driver=driver) as session:
+                list(session.run("RETURN 2 AS n"))
+            with self.session(driver=driver) as session:
+                # same auth token, no re-auth
+                list(session.run("RETURN 3 AS n"))
+            with self.session(driver=driver, auth=self._auth1) as session:
+                # back to original auth token, re-auth required
+                list(session.run("RETURN 4 AS n"))
+        self._reader.done()
+        self.post_test_assertions(self._auth2, self._auth1)
+
+    def test_read_with_switch_routing(self):
+        self.start_server(
+            self._router,
+            self.script_fn_with_features("router_user_switch.script")
+        )
+        self.start_server(
+            self._reader,
+            self.script_fn_with_features("reader_user_switch.script")
+        )
+        with self.driver(routing=True) as driver:
+            with self.session(driver=driver) as session:
+                list(session.run("RETURN 1 AS n"))
+            with self.session(driver=driver, auth=self._auth2) as session:
+                list(session.run("RETURN 2 AS n"))
+            with self.session(driver=driver, auth=self._auth2) as session:
+                # same auth token, no re-auth
+                list(session.run("RETURN 3 AS n"))
+            with self.session(driver=driver, auth=self._auth1) as session:
+                # back to original auth token, re-auth required
+                list(session.run("RETURN 4 AS n"))
+        self._reader.done()
+        self.post_test_assertions(self._auth1, self._auth2)
+
+    def test_read_with_switch_inverse_routing(self):
+        self.start_server(
+            self._router,
+            self.script_fn_with_features(
+                self.inverted_script("router_user_switch.script")
+            )
+        )
+        self.start_server(
+            self._reader,
+            self.script_fn_with_features(
+                self.inverted_script("reader_user_switch.script")
+            )
+        )
+        with self.driver(routing=True, auth=self._auth2) as driver:
+            with self.session(driver=driver, auth=self._auth1) as session:
+                list(session.run("RETURN 1 AS n"))
+            with self.session(driver=driver) as session:
+                list(session.run("RETURN 2 AS n"))
+            with self.session(driver=driver) as session:
+                # same auth token, no re-auth
+                list(session.run("RETURN 3 AS n"))
+            with self.session(driver=driver, auth=self._auth1) as session:
+                # back to original auth token, re-auth required
+                list(session.run("RETURN 4 AS n"))
+        self._reader.done()
+        self.post_test_assertions(self._auth2, self._auth1)
+
+
+class TestUserSwitchingV5x0BackwardsCompatibility(TestUserSwitchingV5x1):
+
+    required_features = (types.Feature.BOLT_5_0,
+                         types.Feature.API_SESSION_AUTH_CONFIG)
```

Comment:
As of now, an additional flag is needed for the backwards compatibility.
@AndyHeap-NeoTech will let us know if this changes.